### PR TITLE
DM: support inject at dml when open GTID

### PR DESCRIPTION
**Pull request by WizardXiao**
### What problem does this PR solve?

Issue Number: ref #4260

### What is changed and how it works?
1. As inject handle will add the original dml to the last event to be inject,  we check weather dml was executed will not check position but gtid and suffix when open gitd, but the last event's suffix will set to be 0, and it will be ignore.
2. Change the location saved table checkpoint or global checkpoint to be a cloned location which suffix is 0, so when check the last event's location will not ignore it, because it will be equal and dml will execute.
3. Change the compare when flush checkpoint to just use compare position, because there is no suffix and the gtid may be equal when handle error.

### Check List 

Tests 

 - Unit test
 - Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: dm/syncer/err-operator/operator.go [hunk 1]
```diff
@@ -177,6 +177,8 @@ func (h *Holder) MatchAndApply(startLocation, endLocation binlog.Location, curre
 	h.mu.Lock()
 	defer h.mu.Unlock()
 
+	h.logger.Info("try to match and apply a operator", zap.Stringer("startlocation", startLocation), zap.Stringer("endlocation", endLocation), zap.Any("currentEvent", currentEvent))
```
Comment (buchuitoudegou) [inline dm/syncer/err-operator/operator.go:180]:
```suggestion
	h.logger.Info("try to match and apply an operator", zap.Stringer("startlocation", startLocation), zap.Stringer("endlocation", endLocation), zap.Any("currentEvent", currentEvent))
```

File: dm/syncer/err-operator/operator.go [hunk 2]
```diff
@@ -215,7 +217,7 @@ func (h *Holder) MatchAndApply(startLocation, endLocation binlog.Location, curre
 		}
 	}
 
-	h.logger.Info("match and apply a operator", zap.Stringer("startlocation", startLocation), zap.Stringer("endlocation", endLocation), zap.Stringer("operator", operator))
+	h.logger.Info("match and apply a operator", zap.Stringer("startlocation", startLocation), zap.Stringer("endlocation", endLocation), zap.Any("currentEvent", currentEvent), zap.Stringer("operator", operator))
```
Comment (buchuitoudegou) [inline dm/syncer/err-operator/operator.go:220]:
```suggestion
	h.logger.Info("match and apply an operator", zap.Stringer("startlocation", startLocation), zap.Stringer("endlocation", endLocation), zap.Any("currentEvent", currentEvent), zap.Stringer("operator", operator))
```

**Review by buchuitoudegou — COMMENTED** [review]
typos😀

File: dm/pkg/binlog/position.go [hunk 3]
```diff
@@ -274,6 +274,13 @@ func (l Location) Clone() Location {
 	return l.CloneWithFlavor("")
 }
 
+// Clone clones a same location, but reset suffix to 0.
```
Comment (Ehco1996) [inline dm/pkg/binlog/position.go:277]: please left a todo location whit suffix is hadr do understand, let's refine this logic later

File: dm/syncer/checkpoint.go [hunk 4]
```diff
@@ -154,8 +154,8 @@ func (b *binlogPoint) outOfDate() bool {
 func (b *binlogPoint) outOfDateBy(pos binlog.Location) bool {
 	b.RLock()
 	defer b.RUnlock()
-
-	return binlog.CompareLocation(pos, b.flushedPoint.location, b.enableGTID) > 0
+	// locations in savedPoint have no suffix and may have same giid like handle-error, we can use position to check them
+	return binlog.CompareLocation(pos, b.flushedPoint.location, false) > 0
```
Comment (Ehco1996) [inline dm/syncer/checkpoint.go:158]: seems change the logic in `binlog.CompareLocation` is more reasonable ?, hardcode to enable gtid to false is too tricky

File: dm/pkg/binlog/position.go (hunk 3 above)
Comment (WizardXiao) [reply dm/pkg/binlog/position.go:277]: yes, suffix logic is hard to understand and it is used in compare location too much. i think i will refine it in other pr.

File: dm/syncer/checkpoint.go (hunk 4 above)
Comment (WizardXiao) [reply dm/syncer/checkpoint.go:158]: `binlog.CompareLocation` is called in many places and i think we may fix it when we refine suffix. this is `outOfDateBy` is used to check between table checkpoint and global checkpoint for flush, this kind of point should have right position, so  we can use position to check.

Comment (Ehco1996) [reply dm/syncer/checkpoint.go:158]: >  add more tests about inject

will you doing this in thie pr or, is this pr ready for review?

**Comment by WizardXiao** [comment]
need fix inject command when it was called in sharding.

File: dm/syncer/checkpoint.go (hunk 4 above)
Comment (WizardXiao) [reply dm/syncer/checkpoint.go:158]: not yet. after some test, it will be more complex when use sharding ddl.

**Pull request closed without merging**